### DISCUSSION TITLE
dataprovider filter with days calculation

### DISCUSSION THREAD

**Discussion by renrom** [Budibase/budibase#9356] [Help] [answered]
I have a form where i want to show a table with a list of items where a sudden (due)date is near today.

For example, want a table where day's to go for now till the duedate in the column is less then 20. The number 20 can be a fixed number

I tried to add a filter on the dataprovider but i can't figure out to let this work.

**Comment by melohagan** [comment]
Hey @renrom 

This looks to be what you are looking for: https://github.com/Budibase/budibase/discussions/3729

**Reply by renrom** [reply]
Hi @melohagan , yeah i have seen that one. but i am not sure yet how to put that as a filter in the datasource.  (i am using the internal database). 

It would be nice to see only the items that are overdue in for example 30 days. But i can't figure out how to put that in the filter of the datasource, as i can't compare the date with the value.  

Still search through all the documentation and this discusions threads to get my hand on budibase.

**Comment by melohagan** [comment, marked as answer]
Hey @renrom 

I understand what you mean. In this case you will need to use a **static** formula. For example:
![Screenshot 2023-01-25 at 13 24 30](https://user-images.githubusercontent.com/101575380/214574625-65f8767d-9aa8-4d86-b48b-bd00f429ebe9.png)

![Screenshot 2023-01-25 at 13 24 48](https://user-images.githubusercontent.com/101575380/214574706-b8290302-dc86-4a96-91d1-f049fa5bb472.png)

```js
const now = new Date();
const expiry = $("DueDate");

if (!expiry) return "Expiry not set";

const dueDate = new Date(expiry);
dueDate.setDate(dueDate.getDate() - 3);

if (dueDate < now) {
	return "Priority";
} else {
	return "Not urgent";
}
```
In your case you would replace the `- 3` with 20.

Then in the design section you can filter on the static formula column:
![Screenshot 2023-01-25 at 13 26 51](https://user-images.githubusercontent.com/101575380/214575158-547d7228-0853-4ad1-bb66-e73c71a732af.png)

---

This is only applicable for the Internal DB. SQL only supports dynamic filters which cannot be filtered on. In that case you could do a custom SQL query and do that date check in the query itself.

**Reply by renrom** [reply]
Hi @melohagan ,

Ok, but i think the problem with the static value is that it is only calculated uppon save? If the record is allready there, then the next day the status cannot change i gues.

**Reply by melohagan** [reply]
Static values should also be created at the point the formula column is created. 
So existing values should get updated when you first add the formula

**Reply by renrom** [reply]
@melohagan 

I think i get you, but i also think that it wont work for me. I have to compare the date with "today" for a given record. So the static value is not usable for me. Let me try to explain: 

- Add an item wich has a dueDate after openingdate. The openingsdate is for example 01/02/2022 and the item is after 12 months not usable anymore. So the duedate is uppon Saving 01/02/2023. So in this examply i have 7 days left. (as it is 25-01-2023 for now). Tomorrrow that value should be 6 and so on. I would like to have a form with a table with only the items that had 30 or less days before overtime. 

hope it makes sence a bit :)

**Reply by renrom** [reply]
@melohagan , i switched over to MySQL as a dataprovider and using custom queries! It gives me (much) more control over presenting the data :). The only little disadvantage is that the external dataproviders don't support attachements or images. but i can live with that.
